pyro/infer/search.py

```diff
@@ -40,3 +42,34 @@ def _traces(self, *args, **kwargs):
         while not self.queue.empty():
             tr = p.get_trace(*args, **kwargs)
             yield (tr, tr.log_prob_sum())
+
+
+class ParallelSearch(TracePosterior):
+    """
+    TODO docs
+    """
+    def __init__(self, model, first_available_dim=None):
+        """
+        TODO docs
+        """
+        if first_available_dim is None:
+            first_available_dim = 0
+        self.first_available_dim = first_available_dim
+        self.model = model
+
+    def _traces(self, *args, **kwargs):
+        """
+        TODO docs
+        """
+        p = poutine.trace(
+            poutine.enum(
+                config_enumerate(self.model, default="parallel"),
+                first_available_dim=self.first_available_dim))
+        tr = p.get_trace(*args, **kwargs)
+        # now compute joint probabilities:
+        # TODO identify all global independence dimensions,
+        # and aggregate over all non-global ones
+        tr.compute_log_prob()
+        log_joints = sum([tr.nodes[name]["log_prob"] for name in tr.nodes
```

Comment:
Maybe the basic datatype should be a dict of factors?
```py
log_factors = {name: site["log_prob"]
               for name, site in tr.nodes.items()
               if site["type"] == "sample"}
return tr, log_factors
```
That way downstream consumers can use it as efficiently as they like, while still being able to resort to
```py
log_joints = sum(log_factors.values())
```